### PR TITLE
Just merge it xd

### PR DESCRIPTION
**Pull request by amarullz**

### CODE REVIEW

**Comment by M1cha** [comment]
why did you close the PR again?

**Review by MLXProjects — DISMISSED** [review]

File: include/aroma/types.h [hunk 1]
```diff
@@ -48,4 +48,25 @@ typedef void * voidp;
...
+}
+static inline float __FMIN(float a, float b){
+  return (a<b)?a:b;
+}
+static inline double __DMAX(double a, double b){
+  return (a>b)?a:b;
+}
+static inline double __DMIN(double a, double b){
+  return (a<b)?a:b;
+}
```
Comment (M1cha) [inline include/aroma/types.h:69]: those names violate the C standard: http://port70.net/~nsz/c/c11/n1570.html#7.1.3
`All identifiers that begin with an underscore and either an uppercase letter or another underscore are always reserved for any use.`

Also, what's the point of e.g. calling it `__MAX` and then have a macro `MAX` alias it? Why not give it that name directly?
Also, `MIN` and `MAX` are defined in many headers and there's a reason non of them are implemented as C functions: so they work with all types.
While the prefixed DMIN/DMAX and FMIN/FMAX might be a good idea, MIN/MAX is not, since most people would expect them to work for all types.

Also, I'd suggest using lowercase names for functions because uppercase function names are generally accepted to be macros which users would expect to support all types.

File: include/aroma/hid/hid.h [hunk 2]
```diff
@@ -153,5 +153,6 @@ byte libaroma_hid_get(LIBAROMA_HID_EVENTP e);
  */
 byte libaroma_hid_config(char * name,char * svalue,dword dvalue);
 
+void libaroma_hid_restart();
```
Comment (M1cha) [inline include/aroma/hid/hid.h:156]: you should enable `-Wstrict-prototypes`. While `void` is optional for C++, in C it has a different, quite dangerous semantic.

File: src/aroma/aroma.c [hunk 3]
```diff
@@ -158,8 +158,11 @@ byte libaroma_start() {
   
 #ifdef LIBAROMA_TTY_KDSETMODE
   ALOGI("KDSETMODE = KD_GRAPHICS");
-  int tty1 = open("/dev/tty1", O_RDWR);
+  char ttydev[32];
+  snprintf(ttydev,32,"/dev/tty%i",LIBAROMA_TTY_KDSETMODE);
```
Comment (M1cha) [inline src/aroma/aroma.c:162]:
```suggestion
  snprintf(ttydev,sizeof(ttydev),"/dev/tty%i",LIBAROMA_TTY_KDSETMODE);
```

File: src/aroma/aroma.c [hunk 4]
```diff
@@ -158,8 +158,11 @@ byte libaroma_start() {
   
 #ifdef LIBAROMA_TTY_KDSETMODE
   ALOGI("KDSETMODE = KD_GRAPHICS");
-  int tty1 = open("/dev/tty1", O_RDWR);
+  char ttydev[32];
```
Comment (M1cha) [inline src/aroma/aroma.c:161]: that length seems arbitrary
- `/dev/tty1` == 9
- max value of `%i` == is platform specific.
- 0-terminator

I suggest doing this so you know the length is always 1
```c
uint8_t value = LIBAROMA_TTY_KDSETMODE ? 1 : 0;
```

With that, you can use a length of 11.

File: src/aroma/aroma.c (hunk 3 above)
Comment (M1cha) [inline src/aroma/aroma.c:162]: while I generally prefer using snprintf, in this case you know the length so you could just use `sprintf`.
Alternatively (what I'd prefer) since the function `libaroma_start` has an error code return value already anyway you could properly check the return value just in case we overlooked something.

A proper check would be:
```c
int rc = snprintf(buf, len, ...);
if (rc < 0 || (size_t)rc >= len) {
    // error
}
```

File: src/aroma/aroma.c [hunk 5]
```diff
@@ -158,8 +158,11 @@ byte libaroma_start() {
...
+  snprintf(ttydev,32,"/dev/tty%i",LIBAROMA_TTY_KDSETMODE);
+  int tty1 = open(ttydev, O_RDWR);
   ioctl(tty1, KDSETMODE, KD_GRAPHICS);
```
Comment (M1cha) [inline src/aroma/aroma.c:164]: I know you didn't change this line in this PR but I just noticed: you forgot to check the return value of `ioctl`.

**Comment by M1cha** [comment]
@amarullz Okay so I just realized that the issues I brought up(and many more) were already present before. So I guess it'd make sense to `just merge` your PR so we have your current work in main and then we can start improving it, agree?

Also, a lot has happened between 2016 and today in terms of coding-standards and how to write safe, qualitative code. Also, I've learned a lot in the meantime :yum:

Would you be fine with us realizing the following things before we start working on new things? I'd be willing to help a lot on this.
- define and apply a `.clang-format` file to gain a unified, automatically applied codestyle across the whole codebase.
- add and fix `-Wstrict-prototypes`
- add and fix `-Wconversion`
- use `const` on pointers wherever possible
- remove typedevs on pointers (like `LIBAROMA_CONFIGP`), because users can't see they're passed by value instead of reference and it doesn't even reduce typing because the `P` simply replaces a `*`
- remove all uses of reserved identifiers
- give all functions lowercase names
- you already wrote function documentation but used `/*` instead of `/**`. convert it to make them show up in doxygen and IDEs.
- properly check the return values of all functions
- remove all commented code
- replace full license texts in source files with `SPDX` identifiers
- add a github actions CI and configure it for use in PRs

What I'd appreciate but is a lot of work and needs your consent:
- change all return values to something where `0` means success and everything else means error
- change the return value type to either of those:
  - `int` (commonly used)
  - `enum` (less common, safer and easier to use since all values are defined)
  - a `struct` with a single `int` or `enum` field. This is not common at all but it should be since it prevents accidentally comparing the value to a constant or even to other ints or enums in case `-Wconversion` or `-Wenum-compare` are disabled

**Comment by amarullz** [comment]
Yea. I agree... The problem with this code is, it didn't progress in so long time, only some place tweaked for some need.
There is a lot of things that need a fix and cleanup. libaroma is used in new and very "OLD" running project (Toll gate system).
That's why I keep my own source personal because I am afraid it will broke "ANDROID" support or if I change it so much
it will broke that old program 😉🥲🙂.

I also use and learn a lot of new coding standard, and used it in my newer c project.

For libaroma, I don't know yet what the best thing I should do. I still need it and library still need backward compatible,
but it was a pain to clean up all mess :)

Maybe I should move it to libaroma2? With whole new API, no backward compatibility. But all optimized code in current
version is usable. And I should limit it just for ui toolkit, not become runtime/vm.

I will try my best...

**Comment by M1cha** [comment]
IMO it's not worth preventing progress due to backwards compatibility.
If a company or person can't or doesn't want to invest the time to migrate to a new version they can just keep using the old version and do minor fixes themselves.

With that being said, yes, every time you break API compatibility you should increase the major version, but the version should not be part of the projects name and does not need a new git repository or anything. If you wanna keep maintain the old version you can simply do that in a release branch.

And while you probably have bigger ideas in mind, I'd be able to do most of the things in my list within a weekend (combined, not each) because I've done them many times in past. So if you want me to help you, just say so.

What I'd ask you to do first, would be to decide on a clangformat-supported codestyle though since that'd change every single line in the repo, making it hard to merge other features once you started working on them.
And since it's your project, you are the only one who gets to decide that :)
I recommend using an unmodified or slightly modified existing style instead of inventing your own but it's your choice.

**Comment by M1cha** [comment]
about versioning: https://semver.org/

**Comment by MLXProjects** [comment]
Here it is my little apport:
I'm pretty sure it's possible to mantain backwards compatibility while rewriting libaroma almost from scratch, the heavy pointers usage is what makes it a lot easier to use :)
If you publish any major rework on this, I compromise to implement backwards compat with anything this currently supports (ALL data types like canvas, controls and graph engine, etc).
PD: I agree into making libaroma just a smaller UI toolkit, with runtime/vm things as a different project. 
Something like what I was doing with libarec (some small library with just the basic recovery functions), the everything-is-a-module concept is great for such projects; plus if we separate runtime from ui toolkit it will be easier to update both.
Also, most times one don't need runtime if just developing ui apps, myself wrote desktop apps thanks to SDL support and they don't need it. Usage vary a lot seen how versatile it is.

**Comment by amarullz** [comment]
I think I can implement new rewriting libaroma version that a lot cleaner. to maintain backward compatibility, I can add wrapper to call new functions.

I have a plan to make calls/naming nicer for every feature. usage likes libaroma_canvas_new or libaroma_listitem_option is something that I want to avoid (its become annoying when the project become bigger).

I will try my best to follow standard.

If you have any suggestion for func/struct/types naming, please tell me.

Something like 
```c
al_canvas * cv = alCanvas(x,y);
```

maybe.. still don't know.... (al -> aroma lib / la -> lib aroma), camel case or snake_case or combination???? 🤔🤔

**Comment by amarullz** [comment]
And maybe I should start using libraries (zlib, png, etc) from external source, and not including in this source.

**Comment by M1cha** [comment]
about external libraries: my suggestion would be to let cmake try this in order:
- use pkgconfig
- check if they are part of the toolchain(no config needed, just link against it)
- build the libraries from  git-submodules.

about naming style: persoally I prefer the [linux kernel](https://www.kernel.org/doc/html/v5.16/process/coding-style.html#typedefs) code style of not using typedefs at all. If you don't like that, you should probably use snakecase names with a `_t` suffix since that's the most common style.

I have to admit that I actually prefer [rust's recommended naming style](https://rust-lang.github.io/api-guidelines/naming.html) though - no matter how unusual that is for C :rofl: 
That'd be snake-case for function and variable names, camel-case for types and e.g. enum-variants, and upper-snakecase for global variables and constants.

**Comment by amarullz** [comment]
Yeah I thinking about using pkg-config. But I dont familiar with cmake 🤣😂. And yes that snake-case with `_t` for type is one that I consider `alcanvas_t * cv = alcanvas_new(x,y);`, but still don't have any good idea for prefix, should it us something like `al_canvas_t` , `alcanvas_t` or `la_canvas_t`.

Because when I start implementing that style, It's hard to change again 😁.

That rust naming not so familiar for me 🤣 

Test sintax
```c
alwin_t * win = alwin_new( ... );
alctl_t * button = albutton_new(win, ...);
// or albtn_new()
alctl_t * list = allist_new(win, ...);

// just idea
alcanvas_t * cv = alwin_get_gb(win);
aldraw_rect(cv, x, y, w, h, ...);
aldraw_text(cv, x, y, "TEST", ...);

alfb_sync();
alcanvas_free(cv);

```

Actually about this naming, we already discuss it long before... That I already annoyed with libaroma naming...

**Comment by M1cha** [comment]
whatever you choose I like it when the function and type names match:
e.g. `alwin_t` and `alwin_new` because that's like a namespace and you'll always know the others when seeing one of them without looking up the others.

**Comment by M1cha** [comment]
and I don't have a preference about the global `al`/`la` prefix. I guess that's up to you :grin:

**Comment by MLXProjects** [comment]
About casing, I feel that snake_case is better for C while camelCase is bestt suited for C++ (because of the classes). Just my thinking, I prefer snake_case.
About prefix, it's quite hard to write a short one without making it easy to confuse with other meanings (al = audio library? la seems better here), but maybe nobody else really thinks that and I'm wrong :P
About pointers/types naming, I love the P ending which means pointer (this made me quickly understand how libaroma works while being new to C code). However, it gets weird if you use words that end with P (i.e. LIBAROMA_GROUPP) but again maybe i'm wrong again.
In summary, current snake_case with la_ prefix and P endings for pointers are nice to me, but not sure what you guys think.
PD: please use tabs instead of two spaces for indentation xd

**Comment by MLXProjects** [comment]
About libraries: yeah, please use submodules to ease the updating task.
About build system: I'm not familiar with cmake neither, but seems a nice alternative. Otherwise just stick to make.
I would like to mantain some script build system just to enable standalone compilation (without needing anything aside from the libaroma repo and a toolchain), while my current implementation is quite broken it proved to work very well and made easy af building aroma apps :) I would be happy to rewrite it too, just for making things easier for new users (as not everybody will need/want the multi-project integration cmake gives)
Libaroma should be treated as a single, small library. This way we make both new and advanced users happy, without needing to sacrificate features/functionality :D

**Comment by amarullz** [comment]
I am a windows guys before 🤣🤣🤣, around 2008-2011 I develop with c in windows smartphone & windows mobile (smarttoolkit + enatives showcase), so my habit with pointer ending with P, using byte, word, dword rather than uint8_t, uint16_t and uint32_t is still too strong in libaroma 😂.

a4-4 Screenshoot-Beta-2
😂😂😂


Now I code more in linux, and have more familiar with how code in open-source & linux community formatted the source.
But I think the postfix P, byte, word, dword will be gone for good reason in next rewrite.

Actually about prefix, I will not use `al_`, but just `al` directly with type of `you name it namespace`, so you will not think
it was audioLibrary module, because next word is so obvious, example: `alcanvas`, `albtn`, `alwin` or if we use la -> `lacanvas`, `lawin_t`, `labtn_t`. But I don't know about that details, If you think it's bad, I will try another way.

About indentation, as @M1cha  mention, I will use clang-format to automatically format the code, so It wont mess anymore, now I we have vscode (using it now), in 2015 I use ultra-edit and format manually everything 😒.

And about build environment, I will learn more about cmake. Or someone maybe can help with build.
I want something that clean and not clobbered. I want every build object not mixed in source.

Features that I will remove:
* Runtime / vm
* Android Recovery only codes, like pause parent, activate cores, etc
* String manipulations
* list & array string
* multi language

Features that I will update / rewrite / change method
* HAL & driver
* stream
* messaging for input device

**Comment by MLXProjects** [comment]
Runtime must be a separate application, just like now but with source outside of main lib; we might keep overridable drivers to make this possible tho.
String manipulations and lang support aren't needed but hashing is quite useful, I've used it a lot for making faster string comparisons (better to use a hash table than a lot of strcmp's).
As for android recovery codes, that's obviously unrelated.
I'll contribute the most I can to the new version if you want me to, I think you've already noticed how much I love this project (just check the new codename I put on my 1.0.1 update :P) but if you want to do things by hand don't worry, I'll still contribute :)
By the way, there it is a little telegram group where I discuss libaroma updates with some involved people, mainly testers but also one dev, it's the pterodon recovery group which you may already know/have heard of (https://t.me/joinchat/FZlpFBdO4Hypy6oVASltqA).

**Comment by M1cha** [comment]
@amarullz clang-format is a configurable formatter with a set of default configs(LLVM, GNU, Google, Chromium, Microsoft, Mozilla, WebKit) that can be used or extended or you can use a fully custom config. My suggestion was to use one of those styles which you like the most. If there's small things you hate in each of them(like tabs vs spaces) you can use the one that matches best and derive from it with small modifications.

@MLXProjects IMO the `P` typedefs don't make any sense and cause problems
- They hide the fact that a variable is a pointer and thus passed by reference(not everybody is familiar with that naming style)
- They do so without any advantage(C has a special, single character syntax for pointers, `*`)
- You can't possibly cover all use cases since there's a lot of permutations of what you can do to variables, a few examples with pointers(I didn't even list arrays):
  - `struct s varname;`
  - `struct s *varname;`
  - `const struct s *varname;`
  - `struct s *const varname;`
  - `const struct s *const varname;`
Among all of the above you can also mix in any type or variable attributes: https://gcc.gnu.org/onlinedocs/gcc/Type-Attributes.html

**Comment by M1cha** [comment]
About byte, word, dword, ... Those are fine as names or typedefs and I don't mind anybody using them but:
- they're all of platform-dependent length and should never be the first choice, prefer to use uintX_t and intX_t instead.
- are they signed or unsigned? the name doesn't tell me that

And just a related piece of information that might be interesting to you: There are platforms where 1 byte != 8 bits :wink:

**Comment by amarullz** [comment]
I use this as vscode setting for formatting, and I like google format without modification 😉
```json
{
	"editor.formatOnPaste": true,
	"editor.formatOnSave": true,
	"editor.formatOnType": true,
	"editor.codeActionsOnSave": {
		"source.fixAll": true
	},
	"clang-format.executable": "clang-format-10",
	"clang-format.style": "google",
	"clang-format.language.c.enable": true,
	"[c]": {
		"editor.defaultFormatter": "xaver.clang-format",
		"editor.wordBasedSuggestions": false,
		"editor.suggest.insertMode": "replace",
		"editor.semanticHighlighting.enabled": true
	}
}
```

**Comment by M1cha** [comment]
If you create a `.clang-format` file in the root of the repo with the following contents you won't need to configure anything and it'll just work with all IDEs and on the console:
```
BasedOnStyle: Google
```

**Comment by amarullz** [comment]
Let me know what do you think about this new source tree:

```text
libaroma
├── .clang-format
├── .gitignore
├── AUTHORS.md
├── LICENSE
├── README.md
├── include
│   ├── aroma.h
│   └── libaroma
│       ├── alcanvas.h
│       ├── alctl.h
│       ├── aldialog.h
│       ├── aldraw.h
│       ├── algfx.h
│       ├── alimage.h
│       ├── alimage_writer.h
│       ├── almotion.h
│       ├── almsg.h
│       ├── alstream.h
│       ├── altext.h
│       ├── alwin.h
│       └── ctl
│           ├── alabel.h
│           ├── albar.h
│           ├── albutton.h
│           ├── alfragment.h
│           ├── alimgbox.h
│           ├── alist_caption.h
│           ├── alist_check.h
│           ├── alist_div.h
│           ├── alist_img.h
│           ├── alist_menu.h
│           ├── alist_option.h
│           ├── alistbox.h
│           ├── alpager.h
│           ├── alprogress.h
│           ├── alscroll.h
│           └── altab.h
├── libaroma.code-workspace
└── src
    ├── arch
    │   ├── arm
    │   │   ├── arch.c
    │   │   ├── arch.h
    │   │   ├── arm_neon_gfx.c
    │   │   └── memset32.S
    │   ├── arm64
    │   │   ├── android_memset.S
    │   │   ├── arch.c
    │   │   ├── arch.h
    │   │   └── arm64_neon_gfx.c
    │   └── x86
    │       ├── android_memset16.S
    │       ├── android_memset32.S
    │       ├── arch.c
    │       ├── arch.h
    │       ├── cache.h
    │       └── sse_gfx.c
    ├── aroma.c
    ├── aroma_config.h
    ├── aroma_graph.c
    ├── aroma_hid.c
    ├── aroma_internal.h
    ├── aroma_ui.c
    ├── aroma_utils.c
    ├── drivers
    │   ├── graph
    │   │   ├── linux_drm
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   ├── linux_fb
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   └── sdl
    │   │       ├── graph_driver.c
    │   │       └── graph_driver.h
    │   └── hid
    │       ├── linux_input
    │       │   ├── hid_driver.c
    │       │   └── hid_driver.h
    │       └── sdl
    │           ├── hid_driver.c
    │           └── hid_driver.h
    ├── libaroma
    │   ├── alcanvas.c
    │   ├── alctl.c
    │   ├── aldialog.c
    │   ├── aldraw.c
    │   ├── algfx.c
    │   ├── alimage.c
    │   ├── alimage_writer.c
    │   ├── almotion.c
    │   ├── almsg.c
    │   ├── alstream.c
    │   ├── altext.c
    │   ├── alwin.c
    │   └── ctl
    │       ├── alabel.c
    │       ├── albar.c
    │       ├── albutton.c
    │       ├── alfragment.c
    │       ├── alimgbox.c
    │       ├── alist_caption.c
    │       ├── alist_check.c
    │       ├── alist_div.c
    │       ├── alist_img.c
    │       ├── alist_menu.c
    │       ├── alist_option.c
    │       ├── alistbox.c
    │       ├── alpager.c
    │       ├── alprogress.c
    │       ├── alscroll.c
    │       └── altab.c
    └── platform
        └── linux
            ├── platform.c
            └── platform.h

20 directories, 96 files

```

**Comment by MLXProjects** [comment]
about src/: 

- why not changing aroma_* to just a*?
- it would be better to define wether to use aroma or libaroma for folders, preferredly I prefer aroma (src/aroma as the current version)
- what about putting platform/arch folders inside of src/aroma?

about src/arch/:

- just put `<arch>.c/.h`, e.g. arm.c, arm.h, i386.c and so on

about libaroma/ctl inside of src and include:

- the alabel/alist file names should have two L's so what about just using `la` instead of `al` everywhere? not sure tho

about src/platform:

- just like src/arch/ use `<platform>.c/.h` directly, to avoid unneeded use of subfolders which will just contain one/two files.

**Comment by amarullz** [comment]
For platform, It can use `/platform/platform_linux.h` and `platform_linux.c` directly, because it just a configuration to set some functions that different in other os, like `file_exists`, `gettime`, etc, we can include all in 1 file actually.

But for `arch`, We can't just use 1 file, because there is features that need to implemented differently per arch, and not everything in C, like memset.S, gfx implementation (vector alpha blending, blit ops, dithering etc) with neon/sse.

The naming still subject to change, previously the folder is not `/src/libaroma` `/include/libaroma`, but `/src/lib` `/include/lib` just like curl (all library in `libs` all program in `src`), it's ok just change it to `aroma` i think.

I explain about source file naming: I use `aroma_xxx.c` for non-library calls files, and `alXXXX.c` for library/publics api 🤔🤔 (I confused myself).

I will try change it with `la_`, let me see again.

Here:
```text
libaroma
├── .clang-format
├── .gitignore
├── AUTHORS.md
├── LICENSE
├── README.md
├── include
│   ├── aroma.h
│   └── aroma
│       ├── la_canvas.h
│       ├── la_ctl.h
│       ├── la_dialog.h
│       ├── la_draw.h
│       ├── la_gfx.h
│       ├── la_image.h
│       ├── la_image_writer.h
│       ├── la_motion.h
│       ├── la_msg.h
│       ├── la_stream.h
│       ├── la_text.h
│       ├── la_win.h
│       └── ctl
│           ├── la_label.h
│           ├── la_bar.h
│           ├── la_button.h
│           ├── la_fragment.h
│           ├── la_imgbox.h
│           ├── la_list_caption.h
│           ├── la_list_check.h
│           ├── la_list_div.h
│           ├── la_list_img.h
│           ├── la_list_menu.h
│           ├── la_list_option.h
│           ├── la_listbox.h
│           ├── la_pager.h
│           ├── la_progress.h
│           ├── la_scroll.h
│           └── la_tab.h
├── libaroma.code-workspace
└── src
    ├── arch
    │   ├── arm
    │   │   ├── arm_neon_gfx.c
    │   │   └── memset32.S
    │   ├── arm64
    │   │   ├── android_memset.S
    │   │   └── arm64_neon_gfx.c
    │   └── x86
    │       ├── android_memset16.S
    │       ├── android_memset32.S
    │       ├── cache.h
    │       └── sse_gfx.c
    ├── aroma.c
    ├── aroma_config.h
    ├── aroma_graph.c
    ├── aroma_hid.c
    ├── aroma_internal.h
    ├── aroma_ui.c
    ├── aroma_utils.c
    ├── drivers
    │   ├── graph
    │   │   ├── linux_drm
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   ├── linux_fb
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   └── sdl
    │   │       ├── graph_driver.c
    │   │       └── graph_driver.h
    │   └── hid
    │       ├── linux_input
    │       │   ├── hid_driver.c
    │       │   └── hid_driver.h
    │       └── sdl
    │           ├── hid_driver.c
    │           └── hid_driver.h
    ├── aroma
    │   ├── la_canvas.c
    │   ├── la_ctl.c
    │   ├── la_dialog.c
    │   ├── la_draw.c
    │   ├── la_gfx.c
    │   ├── la_image.c
    │   ├── la_image_writer.c
    │   ├── la_motion.c
    │   ├── la_msg.c
    │   ├── la_stream.c
    │   ├── la_text.c
    │   ├── la_win.c
    │   └── ctl
    │       ├── la_label.c
    │       ├── la_bar.c
    │       ├── la_button.c
    │       ├── la_fragment.c
    │       ├── la_imgbox.c
    │       ├── la_list_caption.c
    │       ├── la_list_check.c
    │       ├── la_list_div.c
    │       ├── la_list_img.c
    │       ├── la_list_menu.c
    │       ├── la_list_option.c
    │       ├── la_listbox.c
    │       ├── la_pager.c
    │       ├── la_progress.c
    │       ├── la_scroll.c
    │       └── la_tab.c
    └── platform
        ├── platform_linux.c
        ├── platform_linux_buildroot.c
        └── platform_qnx.c

```


I think, `aroma_` prefix is better:
```
libaroma
├── .clang-format
├── .gitignore
├── AUTHORS.md
├── LICENSE
├── README.md
├── include
│   ├── aroma.h
│   └── aroma
│       ├── aroma_canvas.h
│       ├── aroma_ctl.h
│       ├── aroma_dialog.h
│       ├── aroma_draw.h
│       ├── aroma_gfx.h
│       ├── aroma_image.h
│       ├── aroma_image_writer.h
│       ├── aroma_motion.h
│       ├── aroma_msg.h
│       ├── aroma_stream.h
│       ├── aroma_text.h
│       ├── aroma_win.h
│       └── ctl
│           ├── aroma_label.h
│           ├── aroma_bar.h
│           ├── aroma_button.h
│           ├── aroma_fragment.h
│           ├── aroma_imgbox.h
│           ├── aroma_list_caption.h
│           ├── aroma_list_check.h
│           ├── aroma_list_div.h
│           ├── aroma_list_img.h
│           ├── aroma_list_menu.h
│           ├── aroma_list_option.h
│           ├── aroma_listbox.h
│           ├── aroma_pager.h
│           ├── aroma_progress.h
│           ├── aroma_scroll.h
│           └── aroma_tab.h
├── libaroma.code-workspace
└── src
    ├── arch
    │   ├── arm
    │   │   ├── arm_neon_gfx.c
    │   │   └── memset32.S
    │   ├── arm64
    │   │   ├── android_memset.S
    │   │   └── arm64_neon_gfx.c
    │   └── x86
    │       ├── android_memset16.S
    │       ├── android_memset32.S
    │       ├── cache.h
    │       └── sse_gfx.c
    ├── aroma.c
    ├── aroma_config.h
    ├── aroma_graph.c
    ├── aroma_hid.c
    ├── aroma_internal.h
    ├── aroma_ui.c
    ├── aroma_utils.c
    ├── drivers
    │   ├── graph
    │   │   ├── linux_drm
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   ├── linux_fb
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   └── sdl
    │   │       ├── graph_driver.c
    │   │       └── graph_driver.h
    │   └── hid
    │       ├── linux_input
    │       │   ├── hid_driver.c
    │       │   └── hid_driver.h
    │       └── sdl
    │           ├── hid_driver.c
    │           └── hid_driver.h
    ├── aroma
    │   ├── aroma_canvas.c
    │   ├── aroma_ctl.c
    │   ├── aroma_dialog.c
    │   ├── aroma_draw.c
    │   ├── aroma_gfx.c
    │   ├── aroma_image.c
    │   ├── aroma_image_writer.c
    │   ├── aroma_motion.c
    │   ├── aroma_msg.c
    │   ├── aroma_stream.c
    │   ├── aroma_text.c
    │   ├── aroma_win.c
    │   └── ctl
    │       ├── aroma_label.c
    │       ├── aroma_bar.c
    │       ├── aroma_button.c
    │       ├── aroma_fragment.c
    │       ├── aroma_imgbox.c
    │       ├── aroma_list_caption.c
    │       ├── aroma_list_check.c
    │       ├── aroma_list_div.c
    │       ├── aroma_list_img.c
    │       ├── aroma_list_menu.c
    │       ├── aroma_list_option.c
    │       ├── aroma_listbox.c
    │       ├── aroma_pager.c
    │       ├── aroma_progress.c
    │       ├── aroma_scroll.c
    │       └── aroma_tab.c
    └── platform
        ├── platform_linux.c
        ├── platform_linux_buildroot.c
        └── platform_qnx.c
```

**Comment by M1cha** [comment]
IMO the headers shouldn't have any prefix since they already are in their own directory.

And as for sources why does there have to be an "aroma" directory? This is the aroma git repository, what else would it be?
I also don't get why some files are in that directory and some aren't. The source files also don't need any prefix because this whole repo is for aroma.

**Comment by amarullz** [comment]
You right, we don't need prefix for filename 😁, its cleaner.

About "aroma" directory, I like it managed in category, it makes me focus when coding it.

The `/src/artos/*` and `/src/artos/ctl/*` actually can be place directly in `/src`, but I don't like it.

```
libaroma
├── .clang-format
├── .gitignore
├── AUTHORS.md
├── LICENSE
├── README.md
├── include
│   ├── aroma.h
│   └── aroma
│       ├── canvas.h
│       ├── ctl.h
│       ├── dialog.h
│       ├── draw.h
│       ├── gfx.h
│       ├── image.h
│       ├── image_writer.h
│       ├── motion.h
│       ├── msg.h
│       ├── stream.h
│       ├── text.h
│       ├── win.h
│       └── ctl
│           ├── label.h
│           ├── bar.h
│           ├── button.h
│           ├── fragment.h
│           ├── imgbox.h
│           ├── list_caption.h
│           ├── list_check.h
│           ├── list_div.h
│           ├── list_img.h
│           ├── list_menu.h
│           ├── list_option.h
│           ├── listbox.h
│           ├── pager.h
│           ├── progress.h
│           ├── scroll.h
│           └── tab.h
├── libaroma.code-workspace
└── src
    ├── arch
    │   ├── arm
    │   │   ├── arm_neon_gfx.c
    │   │   └── memset32.S
    │   ├── arm64
    │   │   ├── android_memset.S
    │   │   └── arm64_neon_gfx.c
    │   └── x86
    │       ├── android_memset16.S
    │       ├── android_memset32.S
    │       ├── cache.h
    │       └── sse_gfx.c
    ├── aroma.c
    ├── aroma_config.h
    ├── aroma_internal.h
    ├── graph.c
    ├── hid.c
    ├── ui.c
    ├── utils.c
    ├── drivers
    │   ├── graph
    │   │   ├── linux_drm
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   ├── linux_fb
    │   │   │   ├── graph_driver.c
    │   │   │   └── graph_driver.h
    │   │   └── sdl
    │   │       ├── graph_driver.c
    │   │       └── graph_driver.h
    │   └── hid
    │       ├── linux_input
    │       │   ├── hid_driver.c
    │       │   └── hid_driver.h
    │       └── sdl
    │           ├── hid_driver.c
    │           └── hid_driver.h
    ├── aroma
    │   ├── canvas.c
    │   ├── ctl.c
    │   ├── dialog.c
    │   ├── draw.c
    │   ├── gfx.c
    │   ├── image.c
    │   ├── image_writer.c
    │   ├── motion.c
    │   ├── msg.c
    │   ├── stream.c
    │   ├── text.c
    │   ├── win.c
    │   └── ctl
    │       ├── label.c
    │       ├── bar.c
    │       ├── button.c
    │       ├── fragment.c
    │       ├── imgbox.c
    │       ├── list_caption.c
    │       ├── list_check.c
    │       ├── list_div.c
    │       ├── list_img.c
    │       ├── list_menu.c
    │       ├── list_option.c
    │       ├── listbox.c
    │       ├── pager.c
    │       ├── progress.c
    │       ├── scroll.c
    │       └── tab.c
    └── platform
        ├── platform_linux.c
        ├── platform_linux_buildroot.c
        └── platform_qnx.c
```

**Comment by MLXProjects** [comment]
It's beautiful now! Just comfortable to my eyes, and makes faster to know what are you working on without needing to read the prefix (also wastes less screen space) :D
It's very nice, let's keep it if @M1cha agrees.
If you ever care about object names collisioning, that shouldn't ever happen because of how build systems handle file compiling (they don't put objs in a single, flat dir but actually use subdirs) .

**Comment by M1cha** [comment]
> About "aroma" directory, I like it managed in category, it makes me focus when coding it.

I'm not sure if I understand how that helps, but It's fine by me. That code structure LGTM.

**Comment by amarullz** [comment]
> > About "aroma" directory, I like it managed in category, it makes me focus when coding it.
> 
> I'm not sure if I understand how that helps, but It's fine by me. That code structure LGTM.

I do alot switching between one source to another, if whole source keep in one directory without any clear separation, it will take me a while to browse the file that I needed. I think @MLXProjects know what I mean (difficult to explain it 😜)

**Comment by MLXProjects** [comment]
Yeah @amarullz, I understand pefectly! I also work on up to three projects at same time and the folder makes it easier for me to identify which project I'm working at.
Let's use the last source tree you've posted then!
PD: Should we communicate in some other way than this github pull? or it's fine lol

**Comment by amarullz** [comment]
@MLXProjects maybe discord or something?, just invite me I will join.

**Comment by M1cha** [comment]
I'd prefer discord over telegram since it's easier to use on PC

**Comment by MLXProjects** [comment]
I'm mainly a Telegram user, but will use a discord server and maybe bridge it to some Telegram group.
https://discord.gg/AKyQE8mM
